File: src/models/posology_model.py
from typing import Optional
from src.utils import request_data, PORT

class PosologyModel:
    @staticmethod
    def get_posologies(patient_id: int, medication_id: int) -> Optional[list]:
        url = f"http://localhost:{PORT}/{patient_id}/medications/{medication_id}/posologies"
        posologies, status = request_data(url, "GET")
        if status == 200:
            return posologies
        return None
    
    @staticmethod
    def add_posology(patient_id: int, medication_id: int, posology_id: int, hour: int, minute: int) -> Optional[dict]:
        url = f"http://localhost:{PORT}/{patient_id}/medications/{medication_id}/posologies"
        hour = max(0, min(hour, 23))
        minute = max(0, min(minute, 59))
        data = {
            "id": posology_id,
            "hour": hour,
            "minute": minute,
            "medication_id": medication_id
        }
        posology, status = request_data(url, "POST", data)
        if status == 201:
            return posology
        return None
    
    @staticmethod
    def update_posology(patient_id: int, medication_id: int, posology_id: int, hour: int, minute: int) -> Optional[dict]:
        url = f"http://localhost:{PORT}/{patient_id}/medications/{medication_id}/posologies"
        hour = max(0, min(hour, 23))
        minute = max(0, min(minute, 59))
        data = {
            "id": posology_id,
            "hour": hour,
            "minute": minute,
            "medication_id": medication_id
        }
        posology, status = request_data(url, "PATCH", data)
        if status is 204:
            return posology
        return None
    
    @staticmethod
    def delete_posology(patient_id: int, medication_id: int, posology_id: int) -> Optional[dict]:
        url = f"http://localhost:{PORT}/{patient_id}/medications/{medication_id}/posologies/{posology_id}"
        response, status = request_data(url, "DELETE")
        if status == 204:
            return response
        return None

File: ipm/ipm-24-25-pr-ctica-2-ac-dc/src/lib/models/posology_model.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:src/model.dart';

class PosologyModel {
  Future<List<Map<String, dynamic>>> getPosologies(int patientId, int medicationId) async {
    final url = Uri.parse("${Model.path}/patients/$patientId/medications/$medicationId/posologies");
    final response = await http.get(url);

    if (response.statusCode != 200) {
      throw Exception("Error al obtener las posologías");
    }

    // Verifica qué datos estás recibiendo
    List<Map<String, dynamic>> posologies = List<Map<String, dynamic>>.from(jsonDecode(response.body));

    return posologies;
  }

  Future<Map<String, dynamic>> addPosology(int patientId, int medicationId, int minute, int hour) async {
    final url = Uri.parse("${Model.path}/patients/$patientId/medications/$medicationId/posologies");
    final response = await http.post(
      url,
      headers: {"Content-Type": "application/json"},
      body: jsonEncode({
        "medication_id": medicationId,
        "minute": minute,
        "hour": hour,
      }),
    );

    if (response.statusCode != 201) {
      throw Exception("Error al agregar posología");
    }

    return jsonDecode(response.body);
  }

  Future<void> deletePosology(int patientId, int medicationId, int posologyId) async {
    final url = Uri.parse("${Model.path}/patients/$patientId/medications/$medicationId/posologies/$posologyId");
    final response = await http.delete(url);

    if (response.statusCode != 204) {
      throw Exception("Error al eliminar la posología");
    }
  }
}
